File: ANDROID/app/src/main/java/com/example/bdjcrusadeinternalappli/LoaningAddNext.java
package com.example.bdjcrusadeinternalappli;

import android.app.Activity;
import android.app.FragmentTransaction;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import com.fasterxml.jackson.databind.ObjectMapper;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class LoaningAddNext extends Activity {

    User user;
    Equipment equipment;
    Date startDate;
    Date endDate;

    EditText startDateText;
    EditText endDateText;
    Context context;
    ObjectMapper mapper;

    Button validation;
    Button back;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        context = this;

        setContentView(R.layout.loaning_add_next);

        mapper = new ObjectMapper();

        OkHttpClient client = new OkHttpClient();

        Request request = new Request.Builder()
                .url("http://192.168.43.110:8080/user/" + getIntent().getIntExtra("idUser", 0))
                .get()
                .build();

        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(@NotNull Call call, @NotNull IOException e) {
                Log.e("LoaningAddNext", "fail", e);
            }

            @Override
            public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {

                user = mapper.readValue(response.body().string(), User.class);

                mapper = new ObjectMapper();

                OkHttpClient client = new OkHttpClient();

                Request request = new Request.Builder()
                        .url("http://192.168.43.110:8080/equipment/" + getIntent().getIntExtra("idEquipment", 0))
                        .get()
                        .build();

                client.newCall(request).enqueue(new Callback() {
                    @Override
                    public void onFailure(@NotNull Call call, @NotNull IOException e) {
                        Log.e("LoaningAddNext", "fail", e);
                    }

                    @Override
                    public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {

                        equipment = mapper.readValue(response.body().string(), Equipment.class);

                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {

                                startDateText = findViewById(R.id.startDateText);

                                startDateText.setOnFocusChangeListener(new View.OnFocusChangeListener() {

                                    public void onFocusChange(View view, boolean hasFocus) {

                                        if (hasFocus) {

                                            DateDialog dialog = new DateDialog(view);
                                            FragmentTransaction fragmentTransaction = getFragmentManager().beginTransaction();
                                            dialog.show(fragmentTransaction, "DatePicker");
                                        }
                                    }
                                });

                                endDateText = findViewById(R.id.endDateText);

                                endDateText.setOnFocusChangeListener(new View.OnFocusChangeListener() {

                                    public void onFocusChange(View view, boolean hasFocus) {

                                        if (hasFocus) {

                                            DateDialog dialog = new DateDialog(view);
                                            FragmentTransaction fragmentTransaction = getFragmentManager().beginTransaction();
                                            dialog.show(fragmentTransaction, "DatePicker");
                                        }
                                    }
                                });

                                back = findViewById(R.id.back);
                                back.setOnClickListener(new View.OnClickListener() {

                                    @Override
                                    public void onClick(View v) {

                                        Intent intent;
                                        intent = new Intent(v.getContext(), LoaningAddEquipment.class);
                                        intent.putExtra("idUser", getIntent().getIntExtra("idUser", 0));
                                        startActivity(intent);
                                    }
                                });

                                validation = findViewById(R.id.validation);
                                validation.setOnClickListener(new View.OnClickListener() {

                                    @Override
                                    public void onClick(View v) {

                                        if (    !startDateText.getText().equals("") &
                                                !endDateText.getText().equals("")) {

                                            mapper = new ObjectMapper();

                                            OkHttpClient client = new OkHttpClient();

                                            startDate = new Date(startDateText.getText().toString());
                                            endDate = new Date(endDateText.getText().toString());

                                            Loaning loaning = new Loaning(user, equipment, startDate, endDate, "no");

                                            String rq = loaning.toStringWithoutId();

                                            MediaType JSON = MediaType.get("application/json; charset=utf-8");

                                            RequestBody body = RequestBody.create(JSON, rq);

                                            Request request = new Request.Builder()
                                                    .url("http://192.168.43.110:8080/loaning")
                                                    .put(body)
                                                    .build();

                                            client.newCall(request).enqueue(new Callback() {
                                                @Override
                                                public void onFailure(@NotNull Call call, @NotNull IOException e) {
                                                    Log.e("LoaningAddNext", "fail", e);
                                                }

                                                @Override
                                                public void onResponse(@NotNull Call call, @NotNull Response response) throws IOException {

                                                    Intent intent;
                                                    intent = new Intent(v.getContext(), LoaningPage.class);
                                                    intent.putExtra("idUser", getIntent().getIntExtra("idUser", 0));
                                                    startActivity(intent);
                                                }
                                            });
                                        }
                                        else {

                                            Toast.makeText(LoaningAddNext.this, "One or all date(s) isn't picked", Toast.LENGTH_LONG).show();
                                            Log.i("LoaningAddNext", "One or all date(s) isn't picked");
                                        }
                                    }
                                });
                            }
                        });
                    }
                });
            }
        });
    }
}

File: ANDROID/app/src/main/java/kotlinClass/LoaningAddNext.kt
package kotlinClass

import android.app.Activity
import android.app.DatePickerDialog
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import com.example.bdjcrusadeinternalappli.*
import com.example.bdjcrusadeinternalappli.R.*
import com.fasterxml.jackson.databind.ObjectMapper
import okhttp3.*
import java.io.IOException
import java.util.*
import okhttp3.MediaType.Companion.toMediaType

class LoaningAddNext : Activity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        var user: User
        var equipment: Equipment
        var startDate: Date = Date()
        var endDate: Date = Date()

        var startDateText: TextView
        var endDateText: TextView

        var setStartDate: Button
        var setEndDate: Button

        val context = this
        var intent = intent

        var mapper = ObjectMapper()

        var rooterService = RooterService()
        var requestService = RequestService()

        val intentUser = intent.getIntExtra("idUser", 0)
        val intentEquipment = intent.getIntExtra("idEquipment", 0)

        var validation: Button
        var back: Button

        setContentView(layout.loaning_add_next)

        requestService.requestBuilderGet("user", intentUser)
                .enqueue(object : Callback {

            override fun onFailure(call: Call, e: IOException) {

                Toast.makeText(this@LoaningAddNext, "Conversation with server fail", Toast.LENGTH_LONG).show()
            }

            override fun onResponse(call: Call, response: Response) {

                user = mapper.readValue(response.body!!.string(), User::class.java)

                requestService.requestBuilderGet("equipment", intentEquipment)
                        .enqueue(object : Callback {

                    override fun onFailure(call: Call, e: IOException) {

                        Toast.makeText(this@LoaningAddNext, "Conversation with server fail", Toast.LENGTH_LONG).show()
                    }

                    override fun onResponse(call: Call, response: Response) {

                        equipment = mapper.readValue(response.body!!.string(), Equipment::class.java)

                        startDateText = findViewById(id.startDateText)
                        endDateText = findViewById(id.endDateText)

                        setStartDate = findViewById(id.setStartDate)
                        setEndDate = findViewById(id.setEndDate)

                        back = findViewById(id.back)
                        validation = findViewById(id.validation)

                        setStartDate.setOnClickListener(View.OnClickListener {

                            val cal = Calendar.getInstance()
                            val year = cal.get(Calendar.YEAR)
                            val month: Int = cal.get(Calendar.MONTH)
                            val day = cal.get(Calendar.DAY_OF_MONTH)

                            var datePickerDialog = DatePickerDialog(this@LoaningAddNext, DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->

                                var MonthOfYear = monthOfYear + 1
                                startDateText.setText("$dayOfMonth/$MonthOfYear/$year")
                                startDate = Date("$MonthOfYear/$dayOfMonth/$year")

                            }, year, month, day)

                            datePickerDialog.show()
                        })

                        setEndDate.setOnClickListener(View.OnClickListener {

                            val cal = Calendar.getInstance()
                            val year = cal.get(Calendar.YEAR)
                            val month: Int = cal.get(Calendar.MONTH)
                            val day = cal.get(Calendar.DAY_OF_MONTH)

                            var datePickerDialog = DatePickerDialog(this@LoaningAddNext, DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->

                                var MonthOfYear = monthOfYear + 1
                                endDateText.setText("$dayOfMonth/$MonthOfYear/$year")
                                endDate = Date("$MonthOfYear/$dayOfMonth/$year")

                            }, year, month, day)

                            datePickerDialog.show()
                        })

                        back.setOnClickListener(View.OnClickListener {

                            rooterService.changeActivity(Intent(context, LoaningAddEquipment::class.java), context, intentUser)
                        })

                        validation.setOnClickListener { v ->
                            if (startDateText.text != "" && endDateText.text != "") {

                                if (testDateSet(startDate, endDate)) {

                                    var temp = startDate
                                    startDate = endDate
                                    endDate = temp
                                }

                                val loaning = Loaning(user, equipment, startDate, endDate, "In Progress")

                                mapper = ObjectMapper()

                                requestService.requestBuilderPut("loaning", loaning.toStringWithoutId())
                                        .enqueue(object : Callback {

                                    override fun onFailure(call: Call, e: IOException) {
                                        Toast.makeText(this@LoaningAddNext, "Conversation with server fail", Toast.LENGTH_LONG).show()
                                    }

                                    @Throws(IOException::class)
                                    override fun onResponse(call: Call, response: Response) {

                                        rooterService.changeActivity(Intent(v.context, LoaningPage::class.java), context, intentUser)
                                    }
                                })
                            } else {

                                Toast.makeText(this@LoaningAddNext, "One or all field(s) is empty", Toast.LENGTH_LONG).show()
                                Log.i("LoaningAddNext", "One or all field(s) is empty")
                            }
                        }
                    }
                })
            }
        })
    }

    fun testDateSet(startDate: Date, endDate: Date): Boolean {

        if (startDate.time > endDate.time) {

            return true
        }

        return false
    }
}